File: app/src/main/java/com/nguyendevs/stkh/Manager/PermissionManager.java
package com.nguyendevs.stkh.Manager;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.widget.Toast;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import com.nguyendevs.stkh.MainActivity;
import com.nguyendevs.stkh.R;

public class PermissionManager {
    private final Context context;
    private static final int REQUEST_RECORD_AUDIO_PERMISSION = 1;
    private static final int REQUEST_READ_STORAGE_PERMISSION = 2;
    private static final int REQUEST_WRITE_STORAGE_PERMISSION = 3;
    private Runnable onWritePermissionGranted; // Callback for file export

    public PermissionManager(Context context) {
        this.context = context;
    }
    public void checkAudioPermission() {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((MainActivity) context, new String[]{Manifest.permission.RECORD_AUDIO}, REQUEST_RECORD_AUDIO_PERMISSION);
        }
    }
    public void checkStoragePermission() {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_MEDIA_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((MainActivity) context, new String[]{Manifest.permission.READ_MEDIA_AUDIO}, REQUEST_READ_STORAGE_PERMISSION);
        }
    }
    public void checkWriteStoragePermission(Runnable onPermissionGranted) {
        this.onWritePermissionGranted = onPermissionGranted;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            if (onWritePermissionGranted != null) {
                onWritePermissionGranted.run();
            }
            return;
        }
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((MainActivity) context, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_WRITE_STORAGE_PERMISSION);
        } else {
            if (onWritePermissionGranted != null) {
                onWritePermissionGranted.run();
            }
        }
    }
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == REQUEST_RECORD_AUDIO_PERMISSION) {
            if (grantResults.length <= 0 || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(context, "Quyền ghi âm bị từ chối!", Toast.LENGTH_LONG).show();
            }
        } else if (requestCode == REQUEST_READ_STORAGE_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                String selectedMode = context.getSharedPreferences("AppPrefs", Context.MODE_PRIVATE).getString("selectedMode", "Offline");
                if (selectedMode.equals("Online")) {
                    ((MainActivity) context).findViewById(R.id.imageView4).performClick();
                }
            } else {
                Toast.makeText(context, "Quyền truy cập bộ nhớ bị từ chối!", Toast.LENGTH_LONG).show();
            }
        } else if (requestCode == REQUEST_WRITE_STORAGE_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if (onWritePermissionGranted != null) {
                    onWritePermissionGranted.run();
                }
            } else {
                Toast.makeText(context, "Quyền ghi bộ nhớ bị từ chối!", Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/com/nguyendevs/stkh/manager/PermissionManager.kt
package com.nguyendevs.stkh.manager

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.nguyendevs.stkh.util.showToast

/**
 * PermissionManager - Quản lý xin quyền runtime.
 * Migration: Java PermissionManager → Kotlin.
 */
class PermissionManager(private val context: Context) {

    private val activity: Activity get() = context as Activity
    private var onWritePermissionGranted: (() -> Unit)? = null

    companion object {
        const val REQUEST_RECORD_AUDIO = 1
        const val REQUEST_READ_STORAGE = 2
        const val REQUEST_WRITE_STORAGE = 3
    }

    fun checkAudioPermission() {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO)
            != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(Manifest.permission.RECORD_AUDIO),
                REQUEST_RECORD_AUDIO
            )
        }
    }

    fun checkStoragePermission() {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_MEDIA_AUDIO)
            != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(Manifest.permission.READ_MEDIA_AUDIO),
                REQUEST_READ_STORAGE
            )
        }
    }

    fun checkWriteStoragePermission(onPermissionGranted: () -> Unit) {
        this.onWritePermissionGranted = onPermissionGranted
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            // Android 10+: không cần quyền WRITE_EXTERNAL_STORAGE
            onPermissionGranted()
            return
        }
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                REQUEST_WRITE_STORAGE
            )
        } else {
            onPermissionGranted()
        }
    }

    fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
        onStorageGranted: (() -> Unit)? = null
    ) {
        when (requestCode) {
            REQUEST_RECORD_AUDIO -> {
                if (grantResults.isEmpty() || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    context.showToast("Quyền ghi âm bị từ chối!")
                }
            }
            REQUEST_READ_STORAGE -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    onStorageGranted?.invoke()
                } else {
                    context.showToast("Quyền truy cập bộ nhớ bị từ chối!")
                }
            }
            REQUEST_WRITE_STORAGE -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    onWritePermissionGranted?.invoke()
                } else {
                    context.showToast("Quyền ghi bộ nhớ bị từ chối!")
                }
            }
        }
    }
}
